### ISSUE TITLE
PropertyNotSetInConstructor on callable passed as optional constructor argument

### ISSUE THREAD

**Issue by DavideBicego** [vimeo/psalm#7035]
For some reason a PropertyNotSetInConstructor is reported on the code below

```php
<?php

declare(strict_types=1);

use function sleep;

final class TestClass
{
    /** @psalm-var callable(positive-int): void */
    private $sleep;

    /** @psalm-param callable(positive-int ): void|null $sleep */
    public function __construct(?callable $sleep = null)
    {
        $this->sleep = $sleep ?? $this->sleep;
    }

    /** @param positive-int $seconds */
    private function sleep(int $seconds): void {
        sleep($seconds);
    }
}
```

**Comment by DavideBicego** [comment]
https://psalm.dev/r/41321a6a77

**Comment by weirdan** [comment]
There's an uninitialized read when `TestClass()` is instantiated with `null`. It can be demonstrated by replacing callable type with Closure and using typed property: https://3v4l.org/bTYU1
